File: library/src/main/java/io/github/anderscheow/validator/rules/common/RegexRule.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import io.github.anderscheow.validator.rules.BaseRule;

public abstract class RegexRule extends BaseRule {

    private String regex;

    public RegexRule(@NonNull String regex) {
        super("Does not match regex rule");
        this.regex = regex;
    }

    public RegexRule(@NonNull String regex, @StringRes int errorRes) {
        super(errorRes);
        this.regex = regex;
    }

    public RegexRule(@NonNull String regex, @NonNull String errorMessage) {
        super(errorMessage);
        this.regex = regex;
    }

    @Override
    public boolean validate(Object value) {
        if (value == null) {
            throw new NullPointerException();
        }

        if (value instanceof String) {
            return ((String) value).matches(regex);
        }

        throw new ClassCastException("Required String value");
    }
}

File: library/src/main/java/io/github/anderscheow/validator/rules/common/RegexRule.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import io.github.anderscheow.validator.rules.Rule

open class RegexRule : Rule {

    private var regex: String

    constructor(regex: String, @StringRes errorRes: Int) : super(errorRes) {
        this.regex = regex
    }

    constructor(regex: String, errorMessage: String) : super(errorMessage) {
        this.regex = regex
    }

    override fun validate(value: String?): Boolean {
        if (value == null) {
            throw NullPointerException()
        } else {
            return value.matches(regex.toRegex())
        }
    }
}

fun regex(regex: String, @StringRes errorRes: Int): RegexRule = RegexRule(regex, errorRes)

fun regex(regex: String, errorMessage: String): RegexRule = RegexRule(regex, errorMessage)